server/chat-plugins/the-studio.ts

```diff
@@ -0,0 +1,613 @@
+		}
+		buf += `</td>`;
+		if (!suggested && rec.userData.avatar) {
+			buf += `<td style="text-align:center;width:110px;background:rgba(255,255,255,0.4);border-radius:15px;">`;
+			const isCustom = rec.userData.avatar.startsWith('#');
+			buf += `<img style="margin-bottom:-38px;" src="https://${Config.routes.client}/sprites/trainers${isCustom ? '-custom' : ''}/${isCustom ? rec.userData.avatar.slice(1) : rec.userData.avatar}.png" width="80" height="80" />`;
+			buf += `<br /><span style="background:rgba(0,0,0,0.5);padding:1.5px 4px;color:white;font-size:7pt;">Recommended by:`;
+			buf += `<br /><strong>${rec.userData.name}</strong></span></td>`;
+		}
+		buf += `</tbody></table>`;
+		buf += `</div>`;
+		return buf;
+	}
+
+	likeRecommendation(artist: string, title: string, liker: User) {
+		const rec = this.get(artist, title);
+		if (!rec) {
+			throw new Chat.ErrorMessage(`The song titled '${title}' by ${artist} isn't recommended.`);
+		}
+		if (!rec.liked) {
+			rec.liked = {ips: [], userids: []};
+		}
+		if (rec.liked.ips.includes(liker.latestIp) || rec.liked.userids.includes(liker.id)) {
+			throw new Chat.ErrorMessage(`You've already liked this recommendation.`);
+		}
+		rec.likes++;
+		rec.liked.ips.push(liker.latestIp);
+		rec.liked.userids.push(liker.id);
+		saveRecommendations();
+	}
+
+	get(artist: string, title: string, submitter?: string, fromSuggestions = false) {
+		let recs = recommendations.saved;
+		if (fromSuggestions) recs = recommendations.suggested;
+		return recs.find(x => toID(x.artist) === toID(artist) &&
+			toID(x.title) === toID(title) &&
+			(submitter ? toID(x.userData.name) === toID(submitter) : true));
+	}
+
+	getIndex(artist: string, title: string, submitter?: string, fromSuggestions = false) {
+		let recs = recommendations.saved;
+		if (fromSuggestions) recs = recommendations.suggested;
+		const index = recs.findIndex(x => toID(x.artist) === toID(artist) &&
+				toID(x.title) === toID(title) &&
+				(submitter ? toID(x.userData.name) === toID(submitter) : true));
+		return index;
+	}
+}
+
+export const LastFM = new LastFMInterface();
+export const Recs = new RecommendationsInterface();
+
+export const commands: ChatCommands = {
+	registerlastfm(target, room, user) {
+		if (!target) return this.parse(`/help registerlastfm`);
+		this.checkChat(target);
+		target = this.filter(target) || '';
+		if (!target) {
+			throw new Chat.ErrorMessage(`The provided account name has phrases that PS doesn't allow.`);
+		}
+		LastFM.validateAccountName(target);
+		this.sendReply(LastFM.addAccountName(user.id, target.trim()));
+	},
+	registerlastfmhelp: [
+		`/registerlastfm [username] - Adds the provided [username] to the last.fm database for scrobbling.`,
+		`Usernames can only be 2-15 characters long, must start with a letter, and can only contain letters, numbers, hyphens, and underscores.`,
+	],
+
+	async lastfm(target, room, user) {
+		this.runBroadcast(true);
+		this.splitTarget(target, true);
```

Comment:
Also gonna deprecate this very soon. Dunno if you wanna handle this here.

Reply:
i dont think this should be deprecated, but if you do it, i'd rather have you rewrite it upon deprecating/refactoring since i need it to work for `this.targetUsername`

Reply:
I'm deprecating targetUsername as well.

Reply:
well I need that

Reply:
To be clear, I was the one who wanted to deprecate `splitTarget` and `targetUsername`.

There's no need for you to worry about that, though. The migration path will come when it comes, and will not lose features.